File: 개인공부/알고리즘 스터디/7주차/MakeBiggest.py
#https://programmers.co.kr/learn/courses/30/lessons/42883

def solution(number, k):
    lstNumber = list(number)
    targetLen = len(number) - k
    answer = lstNumber[:targetLen]
    count = k

    for i in number[targetLen:]:
        for j in range(targetLen - 1):
            if answer[j] < answer[j+1]:
                del answer[j]
                answer.append(i)  # 뒤에있는 숫자 추가
                count = count - 1
                if count == 0:  #빼야할 갯수만큼 다 뺌
                    return "".join(answer)
                break

        if answer[-1] < i:  #answer 맨뒤랑 현재 추가되는 숫자 비교
            answer[-1] = i  #뒤에있는 숫자 추가

    return "".join(answer)

if __name__ == '__main__':
    result = solution("1924",2)
    print(result)

    result = solution("1231234", 3)
    print(result)

    result = solution("4177252841", 4)
    print(result)

File: 개인공부/알고리즘 스터디/7주차/MakeBiggest.java
package PersonnelStudy.week7;

public class MakeBiggest {
    public String solution(String number, int k) {
        int targetLen = number.length() - k;
        StringBuilder answer = new StringBuilder(number.substring(0,targetLen));
        int count = k;

        for(int i=targetLen;i<number.length();i++){
            for(int j=0;j<targetLen-1;j++){
                if(answer.charAt(j) < answer.charAt(j+1)){
                    answer.deleteCharAt(j);
                    answer.append(number.charAt(i));
                    count--;
                    if(count == 0)
                        return answer.toString();
                    break;
                }
            }
            if(answer.charAt(answer.length()-1) < number.charAt(i))
                answer.setCharAt(answer.length()-1, number.charAt(i));
        }

        return answer.toString();
    }

    public static void main(String[] args) {
        String result = new MakeBiggest().solution("1924", 2);
        System.out.println(result);

        result = new MakeBiggest().solution("4177252841", 4);
        System.out.println(result);
    }
}
